charger: add bq2518x variants

JordanYates opened the pull request:
Rename the bq25180 implementation and files to the more generic bq2518x. This charger family contains the bq25180, bq25186, bq25188 and the standalone (non-I2C) bq25185.

The register maps are practically identical, so the driver should be re-used.

JordanYates committed:
charger: bq25180: rename to `bq2518x`

Rename the bq25180 implementation and files to the more generic bq2518x.
This charger family contains the bq25180, bq25186, bq25188 and the
standalone (non-I2C) bq25185.

The register maps are practically identical, so the driver should be
re-used.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
charger: bq2518x variants

Add support for the bq25186 and bq25188 chip variants.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
tests: drivers: build_all: charger: bq2518x variants

Add the `ti,bq25186` and `ti,bq25188`` to the build all test.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
charger: bq2518x: restrict precharge values in devicetree

Instead of text in the description that specifies the valid values,
add the `enum` property so values are validated at compile time.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
charger: bq2518x: more configuration options

Add configuration for the VSYS target regulation voltage.

Explicitly specify the threshold at which charging will resume, instead
of deriving it from the combination of two devicetree properties which
have no compile-time validation.

Simplify the process of initialising the chip by precomputing the
register values and explicitly writing the whole register, instead of
updating multiple fields individually.

Signed-off-by: Jordan Yates <jordan@embeint.com>

rriveramcrus commented:
Hello @JordanYates,

Thank you for your contribution! Some of this PR duplicates the efforts of #95727 and some of this PR introduces new functionality and improvements. If it's alright with you let's try and get #95727 merged first and then we can rebase some of what you've put together and continue this PR.

By the way your feedback is welcome on #95727 as you're probably pretty familiar with the device. 

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/power/supply/gpio-charger.c?h=v6.17
One last point from me is that I always envisioned standalone charger devices sharing a common GPIO charger driver as was done in the linux kernel. I'd be curious about your thoughts on this idea.

JordanYates committed:
charger: bq2518x: configurable battery limits

Make the battery undervoltage and over-discharge limits configurable
from devicetree.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates commented:
Rebased on main and added configuration for the battery under-voltage and over-discharge current limits.

JordanYates commented on dts/bindings/charger/ti,bq2518x-common.yaml:
```diff
@@ -55,3 +55,32 @@ properties:
+      Battery undervoltage lockout threshold (falling). If the voltage on the
+      BAT pin falls below this value, the BAT to SYS path will be disengaged.
+      There is a 150 mV hysterisis before the path will be re-enabled. Default
+      matches the register reset value.
```
The units could be switched to micro, I would argue it hurts readability here though, and the datasheet values are specified in volts or milliamps, not micro. Counting off zeroes is a pain.
<img width="1001" height="355" alt="image" src="https://github.com/user-attachments/assets/51340098-ffbe-4efc-9051-56e7c92db2e4" />
